pkg/action/util/pod_manager.go

```diff
@@ -0,0 +1,127 @@
+	namespace, name, err := podutil.ParsePodDisplayName(displayName)
+	if err != nil {
+		err = fmt.Errorf("Failed to parse Pod display name %s with uid %s: %v", displayName, uuid, err)
+		glog.Errorf(err.Error())
+		return nil, err
+	}
+
+	pod, err := p.getRunningPod(p.podsGetter.Pods(namespace), name, uuid)
+	if err != nil {
+		err = fmt.Errorf("Failed to get Pod by %s/%s: %v", namespace, name, err)
+		glog.Errorf(err.Error())
+		return nil, err
+	}
+
+	return pod, err
+}
+
+// getRunningPod finds the pod which is in Running phase by its name and uid.
+func (p *PodCachedManager) getRunningPod(podClient v1.PodInterface, name, uid string) (*api.Pod, error) {
+	// Try the cached name first. If it exists, it means some action was applied so it got renamed.
+	// If try the original name first, it may find the pod that is in terminating process from the previous action
+	// as the action doesn't check if the termination process finishes.
+
+	// Pod is not found with the name. Try using the cached name because actions
+	// might have been applied to the pod.
+	if cachedName, ok := p.getLatestValue(name); ok {
+		glog.V(3).Infof("Using cached name %s for pod %s", cachedName, name)
+		pod, err := podutil.GetPodInPhase(podClient, cachedName.(string), api.PodRunning)
+		if err != nil || pod != nil {
+			return pod, err
+		}
+	}
+
+	pod, err := podutil.GetPodInPhase(podClient, name, api.PodRunning)
+	if err != nil || pod != nil {
+		return pod, err
+	}
+
+	// Pod is not found by name. Try using uid to search for the pod.
+	// Try using the cached uid first.
+	if cachedUid, ok := p.getLatestValue(uid); ok {
+		glog.V(3).Infof("Using cached uid %s for pod %s(%s)", cachedUid, name, uid)
+		pod, err = podutil.GetPodInPhaseByUid(podClient, cachedUid.(string), api.PodRunning)
+		if err != nil || pod != nil {
+			return pod, err
+		}
+	}
+
+	pod, err = podutil.GetPodInPhaseByUid(podClient, uid, api.PodRunning)
+	if err != nil || pod != nil {
+		return pod, err
+	}
+
+	return nil, fmt.Errorf("Failed to find Pod with name %s and uid %s", name, uid)
+}
+
+// Need to find the latest pod as there might be multiple action applied to the pod.
+// E.g., If there are three actions in a row for the same pod, the third one need to get the pod name after the change made by the second action.
+//       action1: pod-foo (old) => pod-foo-c (new)
+//       action2: pod-foo (old) => pod-foo-c (from cache) => pod-foo-c-c (new)
+//       action3: pod-foo (old) => pod-foo-c (from cache) => pod-foo-c-c (from cache again) => pod-foo-c-c-c (new)
+func (p *PodCachedManager) getLatestValue(key string) (interface{}, bool) {
+	val, ok := p.podCache.Get(key)
+	if !ok {
+		return val, ok
+	}
+
+	// Value found. Continue to getting further value using the value found as the new key
+	// Set the maximum for the iteration as in any case (e.g., with a bug), we don't want to run the loop infinitely.
+	maxIte := 20
```

Comment:
(1) define a constant for the value of `maxIte`;
(2) log a warning or error in case of reaching to the `maxIte`;